### ISSUE TITLE
Code samples are lacking real-world usage

### ISSUE THREAD

**Issue by alastairs** [prometheus-net/prometheus-net#108]
I'm just getting started with prometheus-net and trying to work out how we can best integrate it into our application. The first metric I'm trying to add is one that counts requests to an external service, and I can't work out how this is supposed to be done from a class utilising DI (constructor injection). 

Here's an example:

```csharp
public class LoggingBlobStorageClient : IBlobStorageClient
{
    private readonly ILogger _logger;
    private readonly IBlobStorageClient _wrappedStore;

    public LoggingBlobStorageClient(ILogger logger, IBlobStorageClient underlyingStore)
    {
        _logger = logger?.ForContext<AzureBlobStore>() ??
                  throw new ArgumentNullException(nameof(logger));

        _wrappedStore = underlyingStore ??
                        throw new ArgumentNullException(nameof(underlyingStore));
    }

    public async Task<Stream> RetrieveBlob(BlobId blobId)
    {
        var requestTotal = Metrics.CreateCounter("azure_blob_request_total", "Number of requests made for blobs", "type", "environment")
            .WithLabels("retrieve", Environment.GetEnvironmentVariable("ASPNETCORE_ENVIRONMENT"));

        requestTotal.Inc();

        _logger.Information("<{@Event}> {BlobId}",
            Events.BlobStorage.GetBlobStarted, blobId.ToString());

        var blobStream = await _wrappedStore.RetrieveBlob(blobId);

        _logger.Information("<{@Event}> {BlobId}",
            Events.BlobStorage.GetBlobCompleted, blobId.ToString());

        return blobStream;
    }
}
```

I _assume_ that it's no big deal to call `Metrics.CreateCounter()` multiple times, but my expectation of calling a factory method multiple times is that I will end up with multiple instances of the counter, and that's something of a concern: how do I know I'm updating the correct instance? Or maybe I'll get an exception at runtime if I try this. 

It _feels_ as though I should be creating my metrics up-front and registering them with my IoC container so they can be injected into the classes which need them. Or maybe creating them as static instances, as [recommended by the Java client](https://github.com/prometheus/client_java#registering-metrics). Does the `ICollectorRegistry` play any part in this, or is that internal only?

I'm finding that some labels (e.g. deployment environment in the example above) are better/more easily resolved at startup than when in use, and their value will never change. 

Overall, I feel I've missed something in picking up this library, and am hoping that there might be some more guidance available somewhere that I've missed.

**Comment by sandersaares** [comment]
In practice, there are various different and equally valid ways to accomplish this. It rather depends on the application architecture. The documentation is indeed fairly nonexistent on the architectural practices angle, as doing a good job describing that without needlessly limiting the set of approaches is nontrivial.

In general, it can be fine to do Metric.CreateX() every time. Ideally, you would indeed cache these instances but the library does consider that one might be doing repeated calls and will return the same instance each time.

Both static instances and pre-created instances are valid approaches.

`ICollectorRegistry` is there to define the set of data to be exported by the exporter instances. It allows you to export multiple data sets on different endpoints, for example, by using separate instances. For managing your collectors it is probably not very relevant.

> I'm finding that some labels (e.g. deployment environment in the example above) are better/more easily resolved at startup than when in use, and their value will never change.

Prometheus best practice is to apply such labels in the [Prometheus database configuration](https://prometheus.io/docs/prometheus/latest/configuration/configuration/#static_config) as hardcoded endpoint-specific labels. Exporters should not export label values that never change for one instance of the app.

While maintaining this type of documentation in this repo is impractical, I will add a reference to the prometheus-users discussion group to the readme. That is probably one of the best places to receive such guidance.

**Comment by alastairs** [comment]
@sandersaares Many thanks for your detailed reply, and for the advice and recommendations you provided; I appreciate the effort you put into it.

I eventually found additional information in the Prometheus docs around writing a client library, and then [the Java client README](https://github.com/prometheus/client_java#registering-metrics) which led me to the approach of declaring my metrics as `static readonly` fields. 

FWIW, I think the Java client README is a really good "getting started" sort of guide to integrating with Prometheus. I think this README would benefit from a bit more information, as one of the hardest things I found was flitting between the Prometheus documentation and the documentation for both this client and the Java client to work out how everything mapped together.
